### PR TITLE
Feature/maintenance page

### PR DESCRIPTION
**Pull request by wshands**
Initial proposed static web page for Dockstore in case of maintenance or downtime for other reasons
Page can be viewed by going to http://d1drt4d6faemoz.cloudfront.net/maintenance_static.html
which pulls the content from https://s3.console.aws.amazon.com/s3/buckets/dockstore-ui-down/test/feature/maintenance-page/?region=us-east-1&tab=overview
Addresses part of https://ucsc-cgl.atlassian.net/browse/SEAB-1396

### CODE REVIEW

**Review by denis-yuen — COMMENTED** [review]
General concept looks good but the page itself could look better.

**Comment by coverbeck** [comment]
> General concept looks good but the page itself could look better.

Agreed. Most status pages that I'm aware of tend to be less-imagey... Not sure how to proceed; coming up with UI is hard.

**Comment by wshands** [comment]
@coverbeck @denis-yuen  @david4096 David provided this [link](https://www.google.com/search?q=maintenance+page+example&client=ubuntu&hs=58p&channel=fs&tbm=isch&source=iu&ictx=1&fir=45zOd26GjXzPEM%252C_77WbbhTQQ3RjM%252C_&vet=1&usg=AI4_-kSiqH-sdXx0m72pky3VEVeylISuaw&sa=X&ved=2ahUKEwjvioa-l6PrAhVCqp4KHUeMCiQQ9QEwAXoECAoQHg&biw=1536&bih=750#imgrc=1lQR8QbVpCNkUM) to example downtime pages 

There are only two images; too many?

Also should this copy the content (HTML and images) to the S3 bucket when there is a push to the repository as is done with https://github.com/dockstore/extra-content/blob/develop/.circleci/config.yml and CircleCI?

**Comment by coverbeck** [comment]
> There are only two images; too many?

I meant the background in your example which covers the whole page, not the number of images. I'm used to status pages being more bare bones without wallpaper. Like the [3rd one from your search example](https://www.google.com/imgres?imgurl=https%3A%2F%2F3kllhk1ibq34qk6sp3bhtox1-wpengine.netdna-ssl.com%2Fwp-content%2Fuploads%2F2018%2F04%2Fbetter-error-pages-maintenance-page-3.png&imgrefurl=https%3A%2F%2Fwww.atlassian.com%2Fblog%2Fstatuspage%2Fmaintenance-page-examples&tbnid=A4d7j-AJOYqgiM&vet=12ahUKEwiDnKHF3LbrAhVQhJ4KHbY6BN4QMygDegUIARC8AQ..i&docid=lOcEJszdLlVzSM&w=1023&h=268&q=maintenance%20page%20example&client=ubuntu&ved=2ahUKEwiDnKHF3LbrAhVQhJ4KHbY6BN4QMygDegUIARC8AQ)

That said, this is very subjective, which is why it's hard to come up with something everybody likes. And that search result did turn up a lot of results, maybe most, with images taking up a lot of the page, which means my vision of a status page may not be the norm.

> Also should this copy the content (HTML and images) to the S3 bucket when there is a push to the repository as is done with https://github.com/dockstore/extra-content/blob/develop/.circleci/config.yml and CircleCI?

Definitely. There will be times when we will have an idea how long we will be out, and so instead of "we hope to be back to be soon", it could be "we expect to be back by 10:30GMT" or something like that. 

BTW, I'm not wild about "hope" in "hope to be back soon" -- we should inspire more confidence. But if you have it updating the content, that can always be tweaked.

Also, in addition to copying the content, you will probably either need to invalidate the CF distribution, or we don't use CF and just serve it up from an s3 bucket directly.

**Comment by wshands** [comment]
@coverbeck I decided to use a new repo instead of extra-content because if I did use the extra-content repo I didn't see a good way to separate in the CircleCI config.yml logic a push for extra-content from a push for maintenance page content. Pipeline variables didn't seem sufficient. https://circleci.com/docs/2.0/pipeline-variables/

As a result I'll need a new CircleCI project setup as was done for extra-content, with an AWS user like was setup for the CircleCI project extra-content and access keys. Could I possibly use the same user and keys for the maintenance-downtime-content CircleCI project?
Thanks!

**Comment by coverbeck** [comment]
> As a result I'll need a new CircleCI project setup as was done for extra-content, with an AWS user like was setup for the CircleCI project extra-content and access keys. Could I possibly use the same user and keys for the maintenance-downtime-content CircleCI project?

I think we should probably have separate keys for this; we already have separate keys for the UI and featured-content. My theory is that it reduces the possibility of one job clobbering the wrong bucket.

**Comment by wshands** [comment]
OK @coverbeck then we can utilize the same user and generate a new set of keys for maintenance content project? I don't know what the user name is, so I cannot generate a new set of keys.

**Comment by denis-yuen** [comment]
re: appearance of status page
I'm ok with one of the simpler ones to start with since we don't have a custom image to work with. 
For example, the AHS, Lonestar, or the status page that Charles linked to. I agree that the background colour and the two images is not what I would normally expect

**Comment by coverbeck** [comment]
> OK @coverbeck then we can utilize the same user and generate a new set of keys for maintenance content project? I don't know what the user name is, so I cannot generate a new set of keys.

What account did you create the s3 bucket in? I think in dev, right?

I think the bucket should be in the other account like our other buckets. Not only to be consistent, but also because I think, although I'm not sure, that if want to have Route53 point to a CF distro, it needs to be in the same account. If you have that set up, we can create a user and keys for you to use.

**Review by denis-yuen — APPROVED** [review]
Could be better, might be a request for the UI redesign. Approving approach for now

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                          command: |
+                              echo ${CIRCLE_BRANCH}
+                              if [ "${CIRCLE_BRANCH}" == "develop" ]; then
+                                    echo 'Uploading downtime content from develop branch to staging folder on bucket'
+                                    echo 'export S3_BUCKET_PATH="develop"' >> $BASH_ENV
+                              elif [ "${CIRCLE_BRANCH}" == "master" ]; then
+                                    echo 'Uploading downtime content from master branch to production folder on bucket'
+                                    echo 'export S3_BUCKET_PATH="production"' >> $BASH_ENV
+                              else
+                                    echo 'Uploading downtime content from testing branch'
```
Comment (coverbeck) [inline .circleci/config.yml]: It's really any branch.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+                                    echo 'export S3_BUCKET_PATH="test/$(echo ${CIRCLE_BRANCH})"' >> $BASH_ENV
+                              fi
+                              echo ${S3_BUCKET_PATH}
+                  - aws-s3/copy:
+                          from: '*.html'
+                          to: 's3://${AWS_BUCKET}/${S3_BUCKET_PATH}/'
+                  - aws-s3/copy:
+                          from: '*.png'
```
Comment (coverbeck) [inline .circleci/config.yml]: I'd suggest a sub-folder and just copy everything. If we add a CSS file or a JPG, we need to add a new entry for each file type.

**Review by coverbeck — APPROVED** [review]
Suggestions aren't crucial; I think this will need to be tweaked UX-wise when we turn this on. It's great to have the infrastructure in place.

File: .circleci/config.yml (hunk 1 above)
Comment (wshands) [reply .circleci/config.yml]: fixed comment

File: .circleci/config.yml (hunk 2 above)
Comment (wshands) [reply .circleci/config.yml]: created sub folder 'content' and now copy everything in it.

**Pull request merged**